rfcs/proposed/task_group_dynamic_dependencies/README.md

```diff
@@ -0,0 +1,311 @@
+
+### Extend the semantics and useful lifetime of task_handle
+
+Dynamic tasks graphs order the execution of tasks via dependencies on the 
+completion of other tasks. They are dynamic because the creation of tasks, 
+specification of dependencies between, submission of tasks for scheduling, and
+execution of the tasks may happen concurrently and in various orders. Different
+concrete use cases have different requirements on when new tasks are created 
+and when dependencies between tasks are specified. 
+
+For the sake of discussion, let’s label four points in a task’s lifetime: 
+
+1. pre-submitted
+2. submitted for dependence tracking
+3. executing
+4. completed. 
+
+A pre-submitted task has been allocated but is not yet known to the scheduling
+algorithm and so cannot begin executing. A submitted task is known to the 
+scheduling algorithm and whenever its incoming dependencies (predecessor tasks)
+are complete it may be scheduled for execution. An executing task has started 
+executing its body but is not yet complete. Finally, a completed task has
+executed fully to completion.
+
+In the current specification for `task_group`, the function `task_group::defer`
+already provides a mechanism to separate task creation from submission. 
+`task_group::defer` returns a `tbb::task_handle`, which represents a deferred 
+task. A deferred task is in the pre-submitted state until it is submitted via
+the `task_group::run` or `task_group::run_and_wait` functions. In the current 
+specification of `task_group`, accessing a `task_handle` after it is submitted
+via one of the run functions is undefined behavior. Currently, therefore, a 
+`task_handle` can only represent a pre-submitted task. And currently, any task
+that is run can immediately be scheduled for execution since there is no notion
+of task dependencies for task_group.
+
+The first extension is to expand the semantics and usable lifetime of
+`task_handle` so that remains valid after it is passed to run and it can 
+represent tasks in any state, including submitted, executing, and completed 
+tasks. Similarly, a `task_handle` in the submitted state may represent a task
+that has predecessors that must complete before it can executed, andso passing
+a `task_handle` to `task_group::run` or `task_group::run_and_wait` only makes
+it available for dependency tracking, and does not make it immediately legal to
+execute.
+
+### Add function(s) to set dependencies.
+
+The obvious next extension is to add a mechanism for specifying dependencies
+between tasks. In the most conservative view, it should only be legal to add 
+additional predecessors / in-dependencies to tasks in the pre-submitted state. 
+After a task starts executing or is completed, it doesn’t make sense to add
```

Comment:
I think this is not 100% correct: If a task is already running and we add a new predecessor, we can read this as a suspend/interrupt. Isn't that what you want to add? We don't need it in our code btw.

Reply:
Hmm, I was not intending to support a suspend. But yes, the language "doesn't make sense" is inaccurate since, as you just pointed out, it can make sense. I'll fix this to match my intention and not be overly broad.